mpeg2/src/ts.rs

```diff
@@ -27,30 +30,112 @@ impl AdaptationField {
         if af_length as usize > buf.len() - 1 {
             return Err(DecodeError::new("adaptation field length too long"));
         } else if buf[0] > 0 {
+            let mut n = 2;
             af.discontinuity_indicator = Some(buf[1] & 0x80 != 0);
             af.random_access_indicator = Some(buf[1] & 0x40 != 0);
             af.program_clock_reference_27mhz = if af_length >= 7 && buf[1] & 0x10 != 0 {
                 let base = (buf[2] as u64) << 25 | (buf[3] as u64) << 17 | (buf[4] as u64) << 9 | (buf[5] as u64) << 1 | (buf[6] as u64) >> 7;
                 let ext = ((buf[6] as u64) & 1) << 8 | (buf[7] as u64);
+                n += 6;
                 Some(base * 300 + ext)
             } else {
                 None
             };
+
+            if buf[1] & 0b0000_1000 != 0 {
+                // skip OPCR
+                n += 6;
+            }
+            if buf[1] & 0b0000_0100 != 0 {
+                // skip splice_countdown
+                n += 1;
+            }
+            if buf[1] & 0b0000_0010 != 0 {
+                if n >= buf.len() || buf[n] as usize + n + 1 > buf.len() {
+                    return Err(DecodeError::new("adaptation field length too long"));
+                }
+                let transport_private_data_length = buf[n];
+                af.private_data_types = buf[n..(n + transport_private_data_length as usize)].to_vec();
+                n += transport_private_data_length as usize + 1
+            }
+            if buf[1] & 1 == 1 {
+                if n + 2 > buf.len() {
+                    return Err(DecodeError::new("adaptation field length too long"));
```

Comment:
```suggestion
                    return Err(DecodeError::new("adaptation field too short for adaptation field extension"));
```

I want to make sure that if we see one of these errors, we can find out where it actually came from. Ideally no two error messages are the same.

Reply:
Absolutely! 👍